File: src/main/java/io/trydent/olimpo/bus/Bus.java
package io.trydent.olimpo.bus;

import io.trydent.olimpo.sys.Id;
import io.vertx.core.eventbus.EventBus;

import java.util.Arrays;

public interface Bus extends Runnable {
  static Bus commandBus(final EventBus bus, final Command... commands) {
    return new CommandBus(
      new CommandProcess(
        Id.uuid(),
        bus
      ),
      Arrays.copyOf(commands, commands.length)
    );
  }

  @Override
  default void run() {
    this.listen();
  }

  void listen();
}

final class CommandBus implements Bus {
  private final Process process;
  private final Command[] commands;

  CommandBus(final Process process, final Command[] commands) {
    this.process = process;
    this.commands = commands;
  }

  @Override
  public final void listen() {
    for (var command : commands) {
      command.executedBy(process);
    }
  }
}

File: src/main/java/io/trydent/olimpo/bus/Bus.kt
package io.trydent.olimpo.bus

interface Bus : () -> Unit {
  companion object {
    fun commandBus(process: Process, vararg commands: Command): Bus = CommandBus(process, arrayOf(*commands))
  }
}

internal class CommandBus(
  private val process: Process,
  private val commands: Array<Command>) : Bus {
  override fun invoke() = commands.forEach { it(process) }
}
